File: Lista/Python/Q2.py
class Carro:
    def __init__(self, marca, modelo, ano):
        self.marca = marca
        self.modelo = modelo
        self.ano = ano
        self.velocidade = 0 

    def exibir_detalhes(self):
        print(f'Marca: {self.marca}, Modelo: {self.modelo}, Ano: {self.ano}')

    def acelerar(self, incremento):
        self.velocidade += incremento
        print(f'{self.modelo} acelerou para {self.velocidade} km/h')

    def frear(self, decremento):
        self.velocidade -= decremento
        if self.velocidade < 0:
            self.velocidade = 0
        print(f'{self.modelo} reduziu para {self.velocidade} km/h')

    def exibir_velocidade(self):
        print(f'Velocidade atual do {self.modelo}: {self.velocidade} km/h')

carro1 = Carro('Toyota', 'Corolla', 2020)
carro2 = Carro('Honda', 'Civic', 2018)
carro3 = Carro('Ford', 'Mustang', 2021)

carro1.exibir_detalhes()
carro1.acelerar(30)
carro1.frear(10)
carro1.exibir_velocidade()

carro2.exibir_detalhes()
carro2.acelerar(50)
carro2.frear(30)
carro2.exibir_velocidade()

carro3.exibir_detalhes()
carro3.acelerar(100)
carro3.frear(50)
carro3.exibir_velocidade()

File: Lista/Go/Q2.go
package main

import "fmt"

type Carro struct {
    Marca      string
    Modelo     string
    Ano        int
    Velocidade int
}

func (c Carro) ExibirDetalhes() {
    fmt.Printf("Marca: %s, Modelo: %s, Ano: %d\n", c.Marca, c.Modelo, c.Ano)
}

func (c *Carro) Acelerar(valor int) {
    c.Velocidade += valor
    fmt.Printf("%s %s acelerou para %d km/h\n", c.Marca, c.Modelo, c.Velocidade)
}

func (c *Carro) Frear(valor int) {
    c.Velocidade -= valor
    if c.Velocidade < 0 {
        c.Velocidade = 0
    }
    fmt.Printf("%s %s reduziu a velocidade para %d km/h\n", c.Marca, c.Modelo, c.Velocidade)
}

func (c Carro) ExibirVelocidade() {
    fmt.Printf("Velocidade atual do %s %s: %d km/h\n", c.Marca, c.Modelo, c.Velocidade)
}

func main() {
    carro1 := Carro{
        Marca:      "Toyota",
        Modelo:     "Corolla",
        Ano:        2020,
        Velocidade: 0,
    }

    carro2 := Carro{
        Marca:      "Honda",
        Modelo:     "Civic",
        Ano:        2018,
        Velocidade: 0,
    }

    carro3 := Carro{
        Marca:      "Ford",
        Modelo:     "Mustang",
        Ano:        2021,
        Velocidade: 0,
    }

    carro1.ExibirDetalhes()
    carro1.ExibirVelocidade()

    carro2.ExibirDetalhes()
    carro2.ExibirVelocidade()

    carro3.ExibirDetalhes()
    carro3.ExibirVelocidade()

    carro1.Acelerar(50)
    carro1.Frear(20)
    carro1.ExibirVelocidade()

    carro2.Acelerar(70)
    carro2.Frear(30)
    carro2.ExibirVelocidade()

    carro3.Acelerar(100)
    carro3.Frear(60)
    carro3.ExibirVelocidade()
}